--- frontends/tasm/tasm.c
+++ frontends/tasm/tasm.c
@@ -224,7 +224,9 @@
 /* version message */
 /*@observer@*/ static const char *version_msg[] = {
     PACKAGE_STRING,
+#if !defined(DONT_EMBED_BUILD_METADATA) || defined(OFFICIAL_BUILD)
     "Compiled on " __DATE__ ".",
+#endif
     "Copyright (c) 2001-2010 Peter Johnson and other Yasm developers.",
     "Run yasm --license for licensing overview and summary."
 };
--- frontends/yasm/yasm.c
+++ frontends/yasm/yasm.c
@@ -213,7 +213,9 @@
 /* version message */
 /*@observer@*/ static const char *version_msg[] = {
     PACKAGE_STRING,
+#if !defined(DONT_EMBED_BUILD_METADATA) || defined(OFFICIAL_BUILD)
     "Compiled on " __DATE__ ".",
+#endif
     "Copyright (c) 2001-2011 Peter Johnson and other Yasm developers.",
     "Run yasm --license for licensing overview and summary."
 };
