Index: dtoa.cc
--- dtoa.cc    (old copy)
+++ dtoa.cc    (working copy)
@@ -183,8 +183,12 @@
 #define NO_HEX_FP
 
 #ifndef Long
+#if __LP64__
+#define Long int
+#else
 #define Long long
 #endif
+#endif
 #ifndef ULong
 typedef unsigned Long ULong;
 #endif
@@ -221,7 +225,7 @@ extern void *MALLOC(size_t);
 #ifndef PRIVATE_MEM
 #define PRIVATE_MEM 2304
 #endif
-#define PRIVATE_mem ((PRIVATE_MEM+sizeof(double)-1)/sizeof(double))
+#define PRIVATE_mem ((unsigned)((PRIVATE_MEM+sizeof(double)-1)/sizeof(double)))
 static double private_mem[PRIVATE_mem], *pmem_next = private_mem;
 #endif
 
