commit 2f12ff08cc9215273040893a9b6c2b3fabccfb6b
Author: Adam Barth <abarth@chromium.org>
Date:   Thu Jul 16 18:00:01 2015 -0700

    Cherry-pick build fix from Mojo

diff --git a/services/android/rules.gni b/services/android/rules.gni
index 350b421..c2c286b 100644
--- a/services/android/rules.gni
+++ b/services/android/rules.gni
@@ -44,6 +44,10 @@ template("mojo_android_java_application") {
   action(android_with_manifest_name) {
     script = "${servicess_android_path}/add_manifest_entry.py"
 
+    deps = [
+      ":$android_standalone_name"
+    ]
+
     input = dex_output_path
     inputs = [
       input,
@@ -74,6 +78,10 @@ template("mojo_android_java_application") {
   action(target_name) {
     script = rebase_path("mojo/public/tools/prepend.py", ".", mojo_root)
 
+    deps = [
+      ":$android_with_manifest_name"
+    ]
+
     input = dex_with_manifest_output_path
     inputs = [
       input,
