commit 2f12ff08cc9215273040893a9b6c2b3fabccfb6b
Author: Adam Barth <abarth@chromium.org>
Date:   Thu Jul 16 18:00:01 2015 -0700

    Improve dependencies

diff --git a/sky/build/sky_app.gni b/sky/build/sky_app.gni
index 950bf82..6890cf6 100644
--- a/sky/build/sky_app.gni
+++ b/sky/build/sky_app.gni
@@ -28,7 +28,7 @@ template("sky_app") {
 
     bundle_prefix = target_name
 
-    copy("copy_${bundle_prefix}_bundle") {
+    copy("gen_${bundle_prefix}_bundle") {
       sources = [
         "$target_gen_dir/app.skyx",
       ]
@@ -50,16 +50,20 @@ template("sky_app") {
         "$target_gen_dir/app.skyx",
       ]
       deps = [
-        "//third_party/icu",
-        ":copy_${bundle_prefix}_bundle",
+        ":$skyx_target_name",
+        ":gen_${bundle_prefix}_bundle",
+        "//third_party/icu:icudata",
       ]
 
       if (defined(invoker.bundles)) {
         foreach(bundle, invoker.bundles) {
           bundle_gen_dir = get_label_info(bundle, "target_gen_dir")
+          bundle_dir = get_label_info(bundle, "dir")
           bundle_name = get_label_info(bundle, "name")
           sources += [ "$bundle_gen_dir/${bundle_name}.skyx" ]
-          deps += [ bundle ]
+          deps += [
+            "$bundle_dir:gen_${bundle_name}_bundle",
+          ]
         }
       }
     }
diff --git a/sky/shell/BUILD.gn b/sky/shell/BUILD.gn
index 88c9d68..ae169d3 100644
--- a/sky/shell/BUILD.gn
+++ b/sky/shell/BUILD.gn
@@ -136,7 +136,7 @@ if (is_android) {
       "$root_build_dir/icudtl.dat",
     ]
     deps = [
-      "//third_party/icu",
+      "//third_party/icu:icudata",
     ]
   }
 } else if (is_ios) {
