diff --git a/engine/src/flutter/lib/ui/BUILD.gn b/engine/src/flutter/lib/ui/BUILD.gn
index e1907d5109a..79fb39ca647 100644
--- a/engine/src/flutter/lib/ui/BUILD.gn
+++ b/engine/src/flutter/lib/ui/BUILD.gn
@@ -323,6 +323,7 @@ if (enable_unittests) {
     public_deps = [
       ":ui",
       "//flutter/common/graphics",
+      "//flutter/display_list",
       "//third_party/googletest:gmock",
     ]
   }
diff --git a/engine/src/flutter/lib/ui/painting/display_list_deferred_image_gpu_impeller_unittests.cc b/engine/src/flutter/lib/ui/painting/display_list_deferred_image_gpu_impeller_unittests.cc
index 003d0ae03ed..4159d25af7b 100644
--- a/engine/src/flutter/lib/ui/painting/display_list_deferred_image_gpu_impeller_unittests.cc
+++ b/engine/src/flutter/lib/ui/painting/display_list_deferred_image_gpu_impeller_unittests.cc
@@ -18,20 +18,6 @@ namespace flutter {
 namespace testing {
 namespace {
 
-class MockDlImage : public DlImage {
- public:
-  MOCK_METHOD(DlISize, GetSize, (), (const, override));
-  MOCK_METHOD(sk_sp<SkImage>, skia_image, (), (const, override));
-  MOCK_METHOD(bool, isOpaque, (), (const, override));
-  MOCK_METHOD(bool, isTextureBacked, (), (const, override));
-  MOCK_METHOD(std::shared_ptr<impeller::Texture>,
-              impeller_texture,
-              (),
-              (const, override));
-  MOCK_METHOD(size_t, GetApproximateByteSize, (), (const, override));
-  MOCK_METHOD(bool, isUIThreadSafe, (), (const, override));
-};
-
 void PostTaskSync(fml::RefPtr<fml::TaskRunner> task_runner,
                   std::function<void()> task) {
   fml::AutoResetWaitableEvent latch;
diff --git a/engine/src/flutter/lib/ui/painting/image_encoding_unittests.cc b/engine/src/flutter/lib/ui/painting/image_encoding_unittests.cc
index 9f7adb7284b..71630f9fe99 100644
--- a/engine/src/flutter/lib/ui/painting/image_encoding_unittests.cc
+++ b/engine/src/flutter/lib/ui/painting/image_encoding_unittests.cc
@@ -34,20 +34,6 @@ namespace testing {
 namespace {
 fml::AutoResetWaitableEvent message_latch;
 
-class MockDlImage : public DlImage {
- public:
-  MOCK_METHOD(sk_sp<SkImage>, skia_image, (), (const, override));
-  MOCK_METHOD(std::shared_ptr<impeller::Texture>,
-              impeller_texture,
-              (),
-              (const, override));
-  MOCK_METHOD(bool, isOpaque, (), (const, override));
-  MOCK_METHOD(bool, isTextureBacked, (), (const, override));
-  MOCK_METHOD(bool, isUIThreadSafe, (), (const, override));
-  MOCK_METHOD(DlISize, GetSize, (), (const, override));
-  MOCK_METHOD(size_t, GetApproximateByteSize, (), (const, override));
-};
-
 class MockSyncSwitch {
  public:
   struct Handlers {
diff --git a/engine/src/flutter/lib/ui/painting/testing/mocks.h b/engine/src/flutter/lib/ui/painting/testing/mocks.h
index 11f012eca98..20378c49fa2 100644
--- a/engine/src/flutter/lib/ui/painting/testing/mocks.h
+++ b/engine/src/flutter/lib/ui/painting/testing/mocks.h
@@ -67,6 +67,20 @@ class MockSnapshotDelegate : public SnapshotDelegate {
   std::shared_ptr<MockTextureRegistry> texture_registry_;
 };
 
+class MockDlImage : public DlImage {
+ public:
+  MOCK_METHOD(DlISize, GetSize, (), (const, override));
+  MOCK_METHOD(sk_sp<SkImage>, skia_image, (), (const, override));
+  MOCK_METHOD(bool, isOpaque, (), (const, override));
+  MOCK_METHOD(bool, isTextureBacked, (), (const, override));
+  MOCK_METHOD(std::shared_ptr<impeller::Texture>,
+              impeller_texture,
+              (),
+              (const, override));
+  MOCK_METHOD(size_t, GetApproximateByteSize, (), (const, override));
+  MOCK_METHOD(bool, isUIThreadSafe, (), (const, override));
+};
+
 }  // namespace testing
 }  // namespace flutter

