diff --git a/mojo/converters/input_events/input_events_type_converters.cc b/mojo/converters/input_events/input_events_type_converters.cc
index 6f2495f..e7b0f15 100644
--- a/mojo/converters/input_events/input_events_type_converters.cc
+++ b/mojo/converters/input_events/input_events_type_converters.cc
@@ -283,7 +283,6 @@ scoped_ptr<ui::Event> TypeConverter<scoped_ptr<ui::Event>, EventPtr>::Convert(
         // TODO: last flags isn't right. Need to send changed_flags.
         scoped_ptr<ui::MouseEvent> event(new ui::MouseEvent(
             MojoMouseEventTypeToUIEvent(input), location, screen_location,
-            base::TimeDelta::FromMilliseconds(input->time_stamp),
             ui::EventFlags(input->flags), ui::EventFlags(input->flags)));
         if (event->IsMouseWheelEvent()) {
           // This conversion assumes we're using the mojo meaning of these
